flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -1705,6 +1710,78 @@ mlir::Value IntrinsicLibrary::genConjg(mlir::Type resultType,
       cplx, negImag, /*isImagPart=*/true);
 }
 
+// COUNT
+fir::ExtendedValue
+IntrinsicLibrary::genCount(mlir::Type resultType,
+                           llvm::ArrayRef<fir::ExtendedValue> args) {
+  assert(args.size() == 3);
+
+  // Handle mask argument
+  auto mask = builder.createBox(loc, args[0]);
+  fir::BoxValue maskTmp = mask;
+  auto maskRank = maskTmp.rank();
+
+  assert(maskRank > 0);
+
+  // Handle optional dim argument
+  bool absentDim = isAbsent(args[1]);
+  auto dim = absentDim
+                 ? builder.createIntegerConstant(loc, builder.getIndexType(), 1)
+                 : fir::getBase(args[1]);
+
+  // Handle optional kind argument
+  auto absentKind = isAbsent(args[2]);
+  auto kind = absentKind ? builder.createIntegerConstant(
+                               loc, builder.getIndexType(),
+                               builder.getKindMap().defaultIntegerKind())
+                         : fir::getBase(args[2]);
+
+  if (absentDim || maskRank == 1) {
+    // Result is scalar if no dim argument or mask is rank 1.
+    // So, call specialized Count runtime routine.
+    if (auto definition = kind.getDefiningOp<ConstantOp>()) {
+      // Kind is a constant
+      if (auto cons = definition.getValue().dyn_cast<IntegerAttr>()) {
+        // Kind is a constant integer
+        // Call genCount() and convert result to the desired kind of integer
+        auto iTY = builder.getIntegerType(
+            builder.getKindMap().getIntegerBitsize(cons.getInt()));
+        auto result = Fortran::lower::genCount(builder, loc, mask, dim);
+        auto temp = builder.createTemporary(loc, iTY);
+        auto cast = builder.createConvert(loc, iTY, result);
+        builder.create<fir::StoreOp>(loc, cast, temp);
+        return builder.create<fir::LoadOp>(loc, temp);
```

Comment:
I do not get why a temp needs to be created here. Why doesn't `return builder.createConvert(loc, resultType, result);` suffice ?

Reply:
I did this because I did not see the conversion right away. Same with recomputing the kind above. I guess we optimize away any immediate covnersions whenever possible. But I did come up with an example where we do not optimize away the conversion of count's result:

call bar(count(mask, kind=2))

bar has an implicit interface, so, the type conversion to i16 (from count's i64 result) is preserved in the case above.

Anyway, I will go back to using just return builder.createConvert().